### PR TITLE
upgrade: Process the upgrade of rabbitmq and epmd packages in the right order

### PR DESCRIPTION
**Pull request by jsuchome**

### CODE REVIEW

File: scripts/upgrade_admin_server.sh [hunk 1]
```diff
@@ -100,6 +100,17 @@ n.save"
 
     # Upgrade the distribution non-interactively
     zypper --no-color --releasever 12.3 ref -f
+
+    systemctl stop rabbitmq-server
+    systemctl stop couchdb
+    sleep 2
+    epmd -kill
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply scripts/upgrade_admin_server.sh]: updated

File: scripts/upgrade_admin_server.sh [hunk 2]
```diff
@@ -100,6 +100,19 @@ n.save"
 
     # Upgrade the distribution non-interactively
     zypper --no-color --releasever 12.3 ref -f
+
+    systemctl stop rabbitmq-server
+    systemctl stop couchdb
+    sleep 5
+    systemctl stop epmd
```
Comment (skazi0) [inline scripts/upgrade_admin_server.sh]: I think it's better to do:
```
systemctl stop epmd.socket
systemctl stop epmd.service
```

File: scripts/upgrade_admin_server.sh [hunk 3]
```diff
@@ -100,6 +100,19 @@ n.save"
 
     # Upgrade the distribution non-interactively
     zypper --no-color --releasever 12.3 ref -f
+
+    systemctl stop rabbitmq-server
+    systemctl stop couchdb
```
Comment (skazi0) [inline scripts/upgrade_admin_server.sh]: Do we really need that?

Comment (jsuchome) [reply scripts/upgrade_admin_server.sh]: I think couchdb could start epmd, that's the reason

Comment (skazi0) [reply scripts/upgrade_admin_server.sh]: oh, I didn't know that couchdb is implemented in erlang :)

File: scripts/upgrade_admin_server.sh [hunk 4]
```diff
@@ -100,6 +100,20 @@ n.save"
...
+    sleep 5
+    systemctl stop epmd.socket
+    systemctl stop epmd.service
+    sleep 5
+    killall epmd
```
Comment (skazi0) [inline scripts/upgrade_admin_server.sh]: What if epmd doesn't react to SIGTERM? shall we wait and try with `killall -9 epmd`? This seems to be the way systemd does stopping by default.

Comment (Itxaka) [reply scripts/upgrade_admin_server.sh]: also, shouldn't we check for epmd state before killing it?

File: scripts/upgrade_admin_server.sh [hunk 5]
```diff
@@ -100,6 +100,20 @@ n.save"
 
     # Upgrade the distribution non-interactively
     zypper --no-color --releasever 12.3 ref -f
+
+    systemctl stop rabbitmq-server
+    systemctl stop couchdb
+    sleep 5
```
Comment (Itxaka) [inline scripts/upgrade_admin_server.sh]: why the sleep? I thought systemctl stop would hang until it stops or fails (timeout/failure to stop)

File: scripts/upgrade_admin_server.sh [hunk 6]
```diff
@@ -125,6 +139,8 @@ ListenStream=[::]:4369
 FreeBind=true
 EOF
 
+    systemctl start rabbitmq-server
```
Comment (Itxaka) [inline scripts/upgrade_admin_server.sh]: will this autostart couchdb?

Comment (skazi0) [reply scripts/upgrade_admin_server.sh]: looking at the rpm scripts, couchdb should be (re)started on package upgrade.

File: scripts/upgrade_admin_server.sh (hunk 4 above)
Comment (skazi0) [reply scripts/upgrade_admin_server.sh]: killall is happy to not kill anything :o)

File: scripts/upgrade_admin_server.sh (hunk 6 above)
Comment (skazi0) [inline scripts/upgrade_admin_server.sh]: it seems to be too late for starting rabbitmq. dist-upgrade above includes crowbar packages which try to upload cookbooks and require rabbit to be running.

File: scripts/upgrade_admin_server.sh (hunk 5 above)
Comment (Gsantomaggio) [reply scripts/upgrade_admin_server.sh]: Added it just for testing.

File: scripts/upgrade_admin_server.sh (hunk 6 above)
Comment (Gsantomaggio) [reply scripts/upgrade_admin_server.sh]: We know ::)
This [patch](https://github.com/crowbar/crowbar/blob/1c9008c6d1704361d6d7f5b75d2e061894b805b9/scripts/upgrade_admin_server.sh#L134) does not work correctly.
I think it is missing a `systemctl daemon-reload`

Comment (skazi0) [reply scripts/upgrade_admin_server.sh]: isn't daemon-reload done inside the package scripts? do you know which package changes some unit which doesn't get reloaded?

File: scripts/upgrade_admin_server.sh (hunk 5 above)
Comment (jsuchome) [reply scripts/upgrade_admin_server.sh]: Yeah, this might not be necessary

File: scripts/upgrade_admin_server.sh (hunk 6 above)
Comment (jsuchome) [reply scripts/upgrade_admin_server.sh]: @skazi0 my reason for placing it here was that it is after that `/etc/systemd/system/epmd.socket.d/ports.conf` creation

Comment (skazi0) [reply scripts/upgrade_admin_server.sh]: I think we should move both parts above zypper dup. I'm testing it right now. Maybe we need another daemon-reload after ports.conf are created.

Comment (skazi0) [reply scripts/upgrade_admin_server.sh]: @jsuchome confirmed: we need another `systemctl daemon-reload` after creating the ports.conf file. So the changes needed for this PR are:
1. move ports.conf creation above zypper dup
2. add `systemctl daemon-reload` after ports.conf are created
3. move `systemctl start rabbitmq-server` above zypper dup but below reload from point 2

**Comment by Gsantomaggio** [comment]
good!

**Pull request merged**